[WIP] Fast path when possible for non numeric aggregation

VibhuJawa opened the pull request:
This PR adds fast path when possible for non numeric aggregation

VibhuJawa commented on dask_sql/physical/rel/logical/aggregate.py:
```diff
@@ -61,6 +61,21 @@ def __init__(self, numerical_aggregation, non_numerical_aggregation=None):
+        if pd.api.types.is_numeric_dtype(dtype):
+            return numeric_agg
+
+        # Todo: Add Categorical when support comes to dask-sql
```
Have not added Categorical as thats not supported currently with dask-sql . 

`string` seems to work as expected for these three aggregations.

VibhuJawa commented:
The PR should be ready for an inital review.

ayushdg commented on dask_sql/physical/rel/logical/aggregate.py:
```diff
@@ -61,6 +61,21 @@ def __init__(self, numerical_aggregation, non_numerical_aggregation=None):
         )
 
+    def get_dtype_relevant_aggregation(self, dtype):
+        numeric_agg = self.numerical_aggregation
```
Not a strong preference but might be worth renaming `numeric_aggregation` `numeric_agg` to something that specifies that these aggregations are not specific to `numerical` types only. 
I don't have a good suggestion for alternative names, but I think of them as `supported` or `native` aggregations.

VibhuJawa replied:
Changed it to `native` and `custom` with comit https://github.com/dask-contrib/dask-sql/pull/236/commits/86b3132f28e366524fb33cd499bc0b223307d04d . Let me know if you would prefer another name,

ayushdg commented on dask_sql/physical/rel/logical/aggregate.py:
```diff
@@ -61,6 +61,21 @@ def __init__(self, numerical_aggregation, non_numerical_aggregation=None):
+            return numeric_agg
+
+        # Todo: Add Categorical when support comes to dask-sql
+        if numeric_agg in ["min", "max", "first"]:
```
Since `first` is the same aggregation for both numeric and non-numeric, no special check is needed for this aggregation. Though for clarity it might make sense to include it here, as well as include `count` which falls in the same criteria.

VibhuJawa replied:
Resolved with removing first  (see commit: https://github.com/dask-contrib/dask-sql/pull/236/commits/89822d92dd9436262d4096fcbecbb371207c6bfe) .

Thanks for the review , good catch on this.

ayushdg reviewed (changes requested):
The changes look great and have a huge performance gains for these specific cases.  Added a few comments inline.

It would be great to include some test cases to [test_groupby](https://github.com/dask-contrib/dask-sql/blob/main/tests/integration/test_groupby.py) as well that add code coverage for this case.

VibhuJawa commented:
> The changes look great and have a huge performance gains for these specific cases. Added a few comments inline.


Thanks , have addressed all of your reviews, I think this is ready for another review. 

> It would be great to include some test cases to [test_groupby](https://github.com/dask-contrib/dask-sql/blob/main/tests/integration/test_groupby.py) as well that add code coverage for this case.

I have included test cases for `StringArray` here, https://github.com/dask-contrib/dask-sql/pull/236/commits/ea48c7a08bef1ec216ebc930e86808566a067b57  in `test_compatibility.py` . 


Adding similar tests was not possible due to not being able to test equality with datetime columns due to a pandas issue with  testing equality b/w string and datetime columns. (Raised the issue here: pd.testing https://github.com/pandas-dev/pandas/issues/43707)  .


For test cases corresponding to `cudf/dask_cudf` , i think we should first let https://github.com/dask-contrib/dask-sql/pull/229/files land to prevent any merge conflicts, Happy to figure something out if you differ on this.

VibhuJawa commented on dask_sql/physical/rel/logical/aggregate.py:
```diff
@@ -48,18 +55,42 @@ class AggregationSpecification:
+
+                ## If pandas StringDtype native aggregation works
+                ## and with ObjectDtype and Nulls native aggregation can fail
+                if isinstance(series, dd.Series) and isinstance(
```
Essentially we need to take the  native agg code path for `Min/Max` if the `series` is either `dask_cudf` with `StringDtype` or `dask.DataFrame` with `pd.StringDtype` .   We want to take the custom agg code path if the dtype is an `object` .  



This is because `.min/max` on an object dtype series with nulls in pandas fails while it works for `StringDtype`   in pandas and works well for `cuDF` too. 
 

```python
#### Works
pd.Series(['a',None],dtype='string').min()
```
#### Fails
```
pd.Series(['a',None],dtype='object').min()

```

I wonder if we can have a cleaner combined check in `cuDF` . CC: @galipremsagar , @ayushdg  for thoughts ?
